### ISSUE TITLE
[Feature Request] More flexible implementation of CNOT gates

### ISSUE THREAD

**Issue by SophLin** [latticesurgery-com/liblsqecc#20]
When congestion results from simultaneous CNOT gates, besides reconsidering the allocation of ancilla (https://github.com/latticesurgery-com/liblsqecc/issues/15#issue-1216018964), one can also try flipping the MXX between ancilla and target, and the MZZ between ancilla and control. This requires a small change to the input format as well. One should specify which pair of MXX and MZZ belong to the same CNOT.

See this paper https://arxiv.org/abs/1704.08670
In ZX calculus, the 3 diagrams below all represent CNOT
<img width="1439" alt="Screen Shot 2022-08-16 at 3 49 30 PM" src="https://user-images.githubusercontent.com/20183375/184982539-5817c57e-638b-4bb0-98fe-0932e7feefc3.png">

**Comment by alexandrupaler** [comment]
It is a nice feature request. We could get it very soon implemented.
The main issue might be related to #15, but we can sort that out also pretty fast.

**Comment by gwwatkin** [comment]
Implementing a new kinds of `cx` gates, distinguishing with QASM comment annotations
 * `cx q[0],q[1]; // SplitControl` 
 * `cx q[0],q[1]; // SplitTarget`

Implementing in two steps. The first step should achieve the minimal viable functionality, enough for testing and unblock you on other things. The second step is to obtain the slice-optimal implementation of the CNOT with lattice surgery. Once I get there we can assess, whether to prioritize that or other things
### 1) Prepare routing and connectivity and parsing only using existing LLI
Pre-allocating the patch used by the third wire:

We already perform CNOTs in a very similar way in the [Python interface repo](https://github.com/latticesurgery-com/lattice-surgery-compiler) things to add are:
* Support for both kinds (Current approach is closer to one of the two)
* Path finding to determine a better placement for the mediating patch (the central wire)
* Simplify corrective terms given that the operation is a CNOT (instead of deriving them from `pi/4` rotations.
##### Example
(excluding corrective terms)
![image](https://user-images.githubusercontent.com/36427091/188341642-9d25bbd4-75b7-417f-b7e3-17f781557ad1.png)


### 2) Minimal ZX implmentation with new LLI
Add new LLI, so that a new patch needs not to be initialized and patches can be merged into a single one.
 * Protrude 
 * Explicit `Merge`
 * Explicit `Split`
 * Shrink
Will define later the details
##### Example
(excluding corrective terms)
![image](https://user-images.githubusercontent.com/36427091/188341661-e32b76a9-5c46-4ba4-bc79-cf366d4bf7dd.png)

**Comment by gwwatkin** [comment]
After discussing with @SophLin, it turns out that approach 1 is as efficient as approach 2 because patch initialization and measuring in the Standard or Hadamard bases takes 0 code-distance-proportional time steps.

Will only implement approach 1 since it does not require any new instructions.

**Comment by gwwatkin** [comment]
Implemented most of the functionality in #20. Now have to work out corrective terms

**Comment by gwwatkin** [comment]
### Corrective terms
![File](https://user-images.githubusercontent.com/36427091/189541579-e639e541-6de7-4c3a-a03c-fee22a6750c8.jpg)
